syntax = "proto3";

package uniset3.umessage;

option optimize_for = LITE_RUNTIME;
// option cc_enable_arenas = true;
// ---------------------------------------------------------------------------
import "google/protobuf/any.proto";
import "UniSetTypes.proto";
// ---------------------------------------------------------------------------
enum Priority
{
    mpLow = 0;
    mpMedium = 1;
    mpHigh = 2;
}

message MessageHeader
{
    Priority priority = 1;
    int64 node = 2;
    int64 supplier = 3;
    int64 consumer = 4;
    Timespec ts = 5;
}

message TransportMessage
{
    Priority priority = 1;
    int64 supplier = 2;
    int64 consumer = 3;
    google.protobuf.Any data = 4;
}

message SensorMessage
{
    MessageHeader header = 1;
    int64 id = 2;
    int64 value = 3;
    Timespec sm_ts = 4;
    IOType sensor_type = 5;
    CalibrateInfo ci = 6;
}

message SystemMessage
{
    enum Command
    {
        Unknown = 0;
        StartUp = 1;    /*! начать работу */
        FoldUp = 2;     /*! нет связи с главной станцией */
        Finish = 3;     /*! завершить работу */
        WatchDog = 4;   /*! контроль состояния */
        ReConfiguration = 5; /*! обновились параметры конфигурации */
        NetworkInfo = 6;     /*! обновилась информация о состоянии узлов в сети
                                 поля
                                 data[0] - кто (ID)
                                 data[1] - новое состояние(true - connect,  false - disconnect)
                             */
        LogRotate = 7;    /*! переоткрыть файлы логов */
        TheLastFieldOfCommand = 8;
    };

    MessageHeader header = 1;
    Command cmd = 2;
    repeated int64 data = 3;
}

message TimerMessage
{
    MessageHeader header = 1;
    int64 id = 2;
}

message ConfirmMessage
{
    MessageHeader header = 1;
    int64 sensor_id = 2;      /* ID датчика (события) */
    double sensor_value = 3; /* значение датчика (события) */
    Timespec sensor_ts = 4;   /*!< время срабатывания датчика (события), который квитируем */
    Timespec confirm_ts = 5;  /*!< время прошедшее до момента квитирования */
    bool broadcast = 6;
    /*!
        признак, что сообщение является пересланным.
        (т.е. в БД второй раз сохранять не надо, пересылать второй раз тоже не надо).
    */
    bool forwarded = 7;
}

message TextMessage 
{
    MessageHeader header = 1;
    string txt = 2;
    int32 mtype = 3;
}
